src/url_pattern_helpers.cpp

```diff
@@ -0,0 +1,1226 @@
+           c == '}' || c == '(' || c == ')' || c == '\\';
+  };
+
+  // While index is less than input’s length:
+  for (const auto& c : input) {
+    if (should_escape(c)) {
+      // then append U+005C (\) to the end of result.
+      result.append("\\");
+    }
+
+    // Append c to the end of result.
+    result += c;
+  }
+  // Return result.
+  return result;
+}
+
+namespace {
+constexpr std::array<uint8_t, 256> escape_regexp_table = []() consteval {
+  std::array<uint8_t, 256> out{};
+  for (auto& c : {'.', '+', '*', '?', '^', '$', '{', '}', '(', ')', '[', ']',
+                  '|', '/', '\\'}) {
+    out[c] = 1;
+  }
+  return out;
+}();
+
+constexpr bool should_escape_regexp_char(char c) {
+  return escape_regexp_table[(uint8_t)c];
+}
+}  // namespace
+
+std::string escape_regexp_string(std::string_view input) {
+  // Assert: input is an ASCII string.
+  ADA_ASSERT_TRUE(idna::is_ascii(input));
+  // Let result be the empty string.
+  std::string result{};
+  result.reserve(input.size());
+  for (const auto& c : input) {
+    // TODO: Optimize this even further
+    if (should_escape_regexp_char(c)) {
+      result.append(std::string("\\") + c);
+    } else {
+      result.push_back(c);
+    }
+  }
+  return result;
+}
+
+std::string process_base_url_string(std::string_view input,
+                                    std::string_view type) {
+  // If type is not "pattern" return input.
+  if (type != "pattern") {
+    return std::string(input);
+  }
+  // Return the result of escaping a pattern string given input.
+  return escape_pattern_string(input);
+}
+
+constexpr bool is_absolute_pathname(std::string_view input,
+                                    std::string_view type) noexcept {
+  // If input is the empty string, then return false.
+  if (input.empty()) [[unlikely]] {
+    return false;
+  }
+  // If input[0] is U+002F (/), then return true.
+  if (input.starts_with("/")) return true;
+  // If type is "url", then return false.
+  if (type == "url") return false;
+  // If input’s code point length is less than 2, then return false.
+  if (input.size() < 2) return false;
+  // If input[0] is U+005C (\) and input[1] is U+002F (/), then return true.
+  if (input.starts_with("\\/")) return true;
+  // If input[0] is U+007B ({) and input[1] is U+002F (/), then return true.
+  if (input.starts_with("{/")) return true;
+  // Return false.
+  return false;
+}
+
+template <url_pattern_encoding_callback F>
+tl::expected<std::vector<url_pattern_part>, url_pattern_errors>
+parse_pattern_string(std::string_view input,
+                     url_pattern_compile_component_options& options,
+                     F&& encoding_callback) {
+  ada_log("parse_pattern_string input=", input);
+  // Let parser be a new pattern parser whose encoding callback is encoding
+  // callback and segment wildcard regexp is the result of running generate a
+  // segment wildcard regexp given options.
+  auto parser = url_pattern_parser<F>(
```

Comment:
The parser's scope is limited to this function, this means that the pointers to the token we generate within this function only lives for the duration of the function. We need to make sure that this is not a problem. It should be documented that no pointer to a token escapes this function.